### ISSUE TITLE
[FEATURE] Solve Constraint with Gradient Descent

### ISSUE THREAD

**Issue by Koukyosyumei** [Koukyosyumei/zkFuzz#5]
Gradient descent might be able to solve constraints that SMT solvers cannot solve efficiently.

## Rule

We can convert logical constraints into numerical functions that yield negative values when the constraints are satisfied. Here are the example conversion rules:

```bash
!(a)     => -a + eps
(a < b)  => a - b + eps
(a <= b) => a - b
(a > b)  => b - a + eps
(a >= b) => b - a
(a == b) => abs(a - b)
(a != b) => -abs(a - b) + eps
(a && b) => max(a, b)
(a || b) => min(a, b)
```

Here, eps is the smallest positive value of the type.

## Algorithm

The algorithm iteratively adjusts input values and free variables to minimize the constraint function:

```python
def proofuzz(circuit, set_of_input_output_pairs):
    constraint = build_constraint(circuit, set_of_input_output_pairs)
    free_variables = extract_unconstrained_intermediate_variables(circuit)

    while True:
        input, output = random.choice(set_of_input_output_pairs)
        initialize_values(input, circuit, free_variables)

        for _ in range(MAX_NUM_ITR):
            loss = evaluate_constraint(input, output, free_variables)
            if loss < 0:
                return input, free_variables
            grads = calculate_gradients(loss, free_variables)
            free_variables -= grads
            free_variables = random_mutate(free_variables)  # if necessary

        mutated_input = random_mutate(input)
        output = circuit(mutated_input)
        set_of_input_output_pairs.append((mutated_input, output))
```

## Example

```c
template Add32Bits() {
    signal input a;     
    signal input b;     
    signal tmp;         
    signal output out; 
    // Check if the sum of 'a' and 'b' exceeds the 32-bit limit (0xFFFFFFFF).
    tmp <-- (a + b) >= (0xFFFFFFFF + 1) ? 1 : 0;
    // Ensure that the overflow flag (tmp) is either 0 or 1.
    tmp * (tmp - 1) === 0;
    // If overflow occurs, subtract 0x100000000 (2^32) from the result to simulate wrap-around.
    out <== (a + b) - (tmp * (0xFFFFFFFF + 1));
}
```

## Existing Works

- Chen, Ju, et al. "Jigsaw: Efficient and scalable path constraints fuzzing." 2022 IEEE Symposium on Security and Privacy (SP). IEEE, 2022.
- Chen, Peng, and Hao Chen. "Angora: Efficient fuzzing by principled search." 2018 IEEE Symposium on Security and Privacy (SP). IEEE, 2018.
